ras-events.c

```diff
@@ -814,6 +814,85 @@ static int add_event_handler(struct ras_events *ras, struct tep_handle *pevent,
 	return 0;
 }
 
+static int report_mce_offline(struct trace_seq *s,
+			      struct mce_event *mce,
+			      struct mce_priv *priv)
+{
+	time_t now;
+	struct tm *tm;
+
+	time(&now);
+	tm = localtime(&now);
+
+	if (tm)
+		strftime(mce->timestamp, sizeof(mce->timestamp),
+			 "%Y-%m-%d %H:%M:%S %z", tm);
+	trace_seq_printf(s, "%s,", mce->timestamp);
+
+	if (*mce->bank_name)
+		trace_seq_printf(s, " %s,", mce->bank_name);
+	else
+		trace_seq_printf(s, " bank=%x,", mce->bank);
+
+	if (*mce->mcastatus_msg)
+		trace_seq_printf(s, " mca: %s,", mce->mcastatus_msg);
+
+	if (*mce->mcistatus_msg)
+		trace_seq_printf(s, " mci: %s,", mce->mcistatus_msg);
+
+	if (*mce->mc_location)
+		trace_seq_printf(s, " Locn: %s,", mce->mc_location);
+
+	if (*mce->error_msg)
+		trace_seq_printf(s, " Error Msg: %s\n", mce->error_msg);
+
+	return 0;
+}
+
+int ras_offline_mc_event(struct ras_mc_offline_event *event)
+{
+	int rc = 0;
+	struct trace_seq s;
+	struct mce_event *mce = (struct mce_event *)calloc(1, sizeof(struct mce_event));
+	struct mce_priv *priv = (struct mce_priv *)calloc(1, sizeof(struct mce_priv));
+
+	trace_seq_init(&s);
+
+	if (event->smca) {
+		priv->cputype = CPU_AMD_SMCA;
+		priv->family = event->family;
+		priv->model = event->model;
+	} else {
+		rc = detect_cpu(priv);
+		if (rc)
+			return -EINVAL;
+	}
+
+	mce->status = event->status;
+	mce->bank = event->bank;
+
+	switch (priv->cputype) {
+	case CPU_AMD_SMCA:
+		mce->ipid = event->ipid;
+		if (!mce->ipid || !mce->status) {
+			printf("%s MSR required.\n", mce->ipid ? "Status" : "Ipid");
+			return -EINVAL;
+		}
+		decode_smca_error(mce, priv);
```

Comment:
Need to #ifdef this otherwise it'll fail on non x86 builds.

Reply:
Or perhaps adding a --with-smca flag so it can be disabled for arches that don't have it

Reply:
Hi aristeu,  I don't see decode_smca_error() function exists in ras-events.c. It has been moved in earlier commits.
Could you please pull the latest commits from rasdaemon or update your repo.

Reply:
This comment is for this pull request (106), you might want to update the pull request then.

Reply:
Hi aristeu, apologies for the delayed response and thanks for bringing this up!

> Need to #ifdef this otherwise it'll fail on non x86 builds.

To make sure that I understand correctly, could you please let me know if you're referring to linker errors like below?

/usr/bin/ld: rasdaemon-ras-events.o: in function `ras_offline_mc_event':
/home/avadnaik/kernel/rasdaemon/ras-events.c:881: undefined reference to `decode_smca_error'

I encountered it using --enable-arm flag with configure. (Wasnt aware of the same. Will keep this in mind subsequently)
autoreconf -vfi && ./configure --enable-arm && make


If yes, can we use the existing ifdef HAVE_MCE to work around it? Something like below:

+++ b/ras-events.c 
+#ifdef HAVE_MCE
+static int report_mce_offline(struct trace_seq *s,
+                             struct mce_event *mce,
+                             struct mce_priv *priv)
+{
.....
+}
+
+int ras_offline_mc_event(struct ras_mc_offline_event *event)
+{
+}
+#endif

+++ b/rasdaemon.c
+#ifdef HAVE_MCE
+       if (args.offline) {
+               ras_offline_mc_event(&event);
+               return 0;
+       }
+#endif

I think this will disable the feature on other arches. Please correct me if I am wrong.

Reply:
Hi aristeu, apologies for the delayed response and thanks for bringing this up!

> Need to #ifdef this otherwise it'll fail on non x86 builds.
To make sure that I understand correctly, could you please let me know if you're referring to linker errors like below?

/usr/bin/ld: rasdaemon-ras-events.o: in function ras_offline_mc_event': /home/avadnaik/kernel/rasdaemon/ras-events.c:881: undefined reference to decode_smca_error'

I encountered it using --enable-arm flag with configure. (Wasnt aware of the same. Will keep this in mind subsequently)
autoreconf -vfi && ./configure --enable-arm && make

If yes, can we use the existing ifdef HAVE_MCE to work around it? Something like below:

+++ b/ras-events.c
+#ifdef HAVE_MCE
+static int report_mce_offline(struct trace_seq *s,
+ struct mce_event *mce, struct mce_priv *priv)
+{
.....
+}
+
+int ras_offline_mc_event(struct ras_mc_offline_event *event)
+{
.....
+}
+#endif

+++ b/rasdaemon.c
+#ifdef HAVE_MCE
+if (args.offline) {
+ras_offline_mc_event(&event);
+return 0;
+}
+#endif

I think this should disable the feature on other arches. Please correct me if I am wrong.

Reply:
That works. But some changes I recommend:
- rename ras_offline_mc_event to ras_offline_mca_event, as it's specific to MCA, not MC.
- following renaming, move it to ras-mce-handler.c. With that, you don't need a ifdef around these functions
because ras-mce-handler.c is only built if HAVE_MCE is set

Reply:
> That works. But some changes I recommend:
> 
> * rename ras_offline_mc_event to ras_offline_mca_event, as it's specific to MCA, not MC.
> * following renaming, move it to ras-mce-handler.c. With that, you don't need a ifdef around these functions
>   because ras-mce-handler.c is only built if HAVE_MCE is set

We can do that. But, IIUC, we may still have to use HAVE_MCE ifdef in rasdaemon.c file to prevent undefined reference errors to ras_offline_mca_event() from the file. Some other changes might also be required.
Will incorporate your suggested changes and modify the pull request accordingly. Will reach out if I encounter any issues.

Additionally, when I use the --enable-arm flag with configure, am also noticing the below undefined reference error:
(root@rasdaemon$ autoreconf -vfi && ./configure --enable-arm && make)

/usr/bin/ld: rasdaemon-ras-arm-handler.o: in function `display_raw_data':
/home/avadnaik/kernel/rasdaemon/ras-arm-handler.c:39: undefined reference to `print_le_hex'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:689: rasdaemon] Error 1

Is this a known issue?

Reply:
> We can do that. But, IIUC, we may still have to use HAVE_MCE ifdef in rasdaemon.c file to prevent undefined reference errors to ras_offline_mca_event() from the file. Some other changes might also be required. Will incorporate your suggested changes and modify the pull request accordingly. Will reach out if I encounter any issues.

Sorry, I should have been more clear. Yes, that's what I've meant. You can keep ifdef on rasdaemon.c but moving
(and renaming as _mc_ is supposed to be generic memory check, not necessarily MCA) the implementation to
a more appropriate file that only gets built when MCA is enabled.

> Additionally, when I use the --enable-arm flag with configure, am also noticing the below undefined reference error: (root@rasdaemon$ autoreconf -vfi && ./configure --enable-arm && make)
> 
> /usr/bin/ld: rasdaemon-ras-arm-handler.o: in function `display_raw_data': /home/avadnaik/kernel/rasdaemon/ras-arm-handler.c:39: undefined reference to `print_le_hex' collect2: error: ld returned 1 exit status make[2]: *** [Makefile:689: rasdaemon] Error 1
> 
> Is this a known issue?

I believe it's because it's a configuration intended for building for ARM machines and that should exist in such
environments. To be honest --enable-arm should be removed and automatic when building on aarch64, as it's
its purpose.

Thanks!

Reply:
Sounds good, thanks for the clarification. Will incorporate your feedback and update the pull request.